Feature/journal migrations

apfelfabrik opened the pull request:
OP Issue: https://www.openproject.org/issues/1793. To test, one can do the following:

```
bundle update                                 &&
bundle exec rake db:drop                      &&
git checkout -- Gemfile.lock                  &&
git checkout 034bcd1                          &&
bundle update                                 &&
bundle exec rake db:create db:migrate db:seed
```

Then, remove all work packages, and continue:

```
git checkout -- Gemfile.lock                  &&
git checkout feature/journal-migrations       &&
bundle update                                 &&
bundle exec rake db:migrate
```

~~:exclamation: Depends on https://github.com/opf/openproject/pull/372. No longer depends on https://github.com/opf/openproject/pull/393.~~

apfelfabrik committed:
Adds migration for legacy journals.

Legacy journals need to be transformed to current journals. This
amazing new migration does it.

apfelfabrik committed:
Check migration prerequisites.

apfelfabrik committed:
Adds journal creation to migration.

apfelfabrik committed:
Extend legacy journal migration

The legacy journal migration now includes normalization into the data
tables. This involves combining previous journals (diffs) into the new
normalized schema and extends to changesets, news, messages and work
packages.

apfelfabrik committed:
Import seeds and Gemfile from feature/rails3.

git checkout opf/feature/rails3 -- db/seeds.rb db/seeds Gemfile

apfelfabrik committed:
Introduces new migration dated after previous fixes.

apfelfabrik committed:
Remvoes obsolete migration.

apfelfabrik committed:
Adds support for more journals.

Now also migrated wiki_content_journals as well as time_entry_journals.
Also, wikis now have some custom behaviour for data -> text conversion.

apfelfabrik committed:
Adds support for work package attachments and custom fields.

Those are ignored for now.

apfelfabrik committed:
Adds support for lock_version in wikis.

apfelfabrik committed:
Introroduces order in journal migration.

This is so that attachments and custom fields are migrated when
workpackages are first touched. It satisfies the need for filenames
when connecting a work_package_journal with attachement_journals trough
attachable_journals, filename being a not null requirement on the
latter.

apfelfabrik committed:
Adds support for empty changesets.

When just commenting, e.g., a work_package, one can omit all changes,
leading to an empty changed_data object. This is now supported.

apfelfabrik committed:
Adds support for updates in subsequent runs of the migration.

apfelfabrik committed:
Moves updates closer together.

Updates can happen really late, and can be wrapped into one call to
execute. This saves about ~25% of execution time on subsequent migration
runs.

apfelfabrik committed:
Adds support for attachable_journals.

apfelfabrik committed:
Move migration to a newer timestamp.

apfelfabrik committed:
Fixes some attachment migration bugs.

apfelfabrik committed:
Makes one of the update statements optional.

Keys might have been empty for the update previously.

apfelfabrik committed:
Merge remote-tracking branch 'opf/feature/rails3' into HEAD

apfelfabrik commented:
I have rebased and shortened this PR for easier reviews.

apfelfabrik committed:
Adds support for custom fields.

Almost like attachable journals.

apfelfabrik committed:
Adds entry into CHANGELOG.

ulferts commented on db/migrate/20130903092236_awesome_legacy_journals_migration.rb:
```diff
@@ -0,0 +1,352 @@
+      # remember the combined journal as the previous one for the next iteration.
+      previous_journal = combined_journal
+      previous_journaled_id = journaled_id
+      previous_type = type
```
You could use a struct here to semantically bundle those values.

ulferts commented on db/migrate/20130903092236_awesome_legacy_journals_migration.rb:
```diff
@@ -0,0 +1,352 @@
+  def check_assumptions
+
+    invalid_journals = ActiveRecord::Base.connection.select_values <<-SQL
+      SELECT DISTINCT tmp.id
```
Please add a comment explaining the intend of this sql

ulferts commented:
When applying this migration to a current openproject.org dump some problems arise. So I would like to continue development on this before merging.

ulferts committed:
fixes validation query to make it pass on my machine

ulferts committed:
fixes inserting columns that are stored in the legacy journal but do not exist in the new journal table

This might e.g. occur when a plugin has added columns to the journaled table which is than journaled.

ulferts committed:
adds a struct to semantically group the previous state

ulferts committed:
use instance variables as this is what is intended

ulferts committed:
adds comment for sql statement

ulferts committed:
complete refactoring to ease reading - still a lot to do

ulferts committed:
a whole lot of cleaning up also fixes, sorry

ulferts committed:
restructures reusable migrator concerns into their own module

ulferts committed:
changes the container of attachment journals from Issue to WorkPackage

ulferts committed:
adds copyright notice

ulferts committed:
Merge branch 'feature/rails3' into feature/journal-migrations

ulferts committed:
renaming migration to take care of dependencies

ulferts committed:
extracts migrator concerns

ulferts committed:
extracts DbWorker

ulferts committed:
extracts journal migrator

ulferts committed:
updates copyright

ulferts committed:
whitespace fix

ulferts committed:
fixes module references

ulferts committed:
updates journaled attribute for former planning_element journals

ulferts committed:
fixes migration of planning_element_journals

ulferts committed:
adds down migration

ulferts committed:
splits single migration into multiple

ulferts committed:
cleaning up down migrations

ulferts committed:
fixing journable_type for former issue journals

apfelfabrik committed:
Removes some comments.

You could put those everywhere.

apfelfabrik committed:
Changes some error messages.

I checked, and there is no noting. It was therefore concluded, that
nothing actually happens instead.

apfelfabrik committed:
Shortens the preconditions_met? method.

apfelfabrik committed:
Fixes changelog.

apfelfabrik committed:
Adds a progress bar.

With full decimal percentage display, this is both more responsive
and less noisy in using up screen real estate. Also, we can reuse the
progressbar in db:seeds rake task.

apfelfabrik commented:
I also ran into broken journals and inconsistencies. I deleted some journals that wouldn't migrate, one with which I had multibyte encoding + varchar limit problem (probably due to borked setup on my side) and fixed handling of journals where no changed_data is attached (no idea where those come from, but they're in my legacy data). Also, I had to fix the not-null constraint on page_id w/ wiki content journals.

I had to change multiple unrelated migrations to get to the journal migrations, mostly stuff we either pulled into or out of the core. Those issues are in separate tickets.

I also took the liberty of adding a progressbar w/ ETA. With such long running migrations I believe that to be a good idea. We can also use it for db:seed, which takes a while, too.

apfelfabrik committed:
Adds page_id to be inserted in journals.

This is due to the not-null constraint on page_id in the wiki content
journals.

apfelfabrik committed:
Treats nil changed_data gracefully.

It would fail rather hard previously.

apfelfabrik committed:
Fixes missing sequence update.

apfelfabrik committed:
Merge remote-tracking branch 'opf/feature/rails3' into feature/journal-migrations

Conflicts:
	doc/CHANGELOG.md

apfelfabrik committed:
Limits reset_pk_sequence to postgres.
